File: memoization/python/06_can_construct.py
"""Can construct.

This script implements a solution to the can construct problem using memoization.

Given a string 'target' and a list of strings 'word_bank', return a boolean indicating
whether or not the 'target' can be constructed by concatenating elements of 'word_bank'
list.

You may reuse elements of 'word_bank' as many times as needed.
"""
from typing import List


def can_construct(target: str, word_bank: List[str], memo: dict = None) -> bool:
    """Check if a given target can be constructed with a given word bank.

    Args:
        target (str): String that must be constructed.
        word_bank (List[str]): List of strings that can possibly construct the target.
        memo (dict): A hashmap to cache combinations already calculated.

    Returns:
        bool: Returns True if it can be constructed or False if it cannot.
    """
    if memo is None:
        memo = {}

    if target in memo:
        return memo[target]

    if not target:
        return True

    for word in word_bank:
        if target.startswith(word):
            if can_construct(target[len(word) :], word_bank, memo):
                memo[target] = True
                return True

    memo[target] = False
    return False


if __name__ == "__main__":
    print(can_construct("abcdef", ["ab", "abc", "cd", "def", "abcd"]))  # True
    print(
        can_construct("skateboard", ["bo", "rd", "ate", "t", "ska", "sk", "boar"])
    )  # False
    print(
        can_construct("enterapotentpot", ["a", "p", "ent", "enter", "ot", "o", "t"])
    )  # True
    print(
        can_construct(
            "eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef",
            ["e", "ee", "eee", "eeee", "eeeee", "eeeeee"],
        )
    )  # False

File: memoization/golang/06_can_construct.go
// Can construct.

// Implementation of a solution to the can construct problem using memoization.

// Given a string 'target' and a list of strings 'wordBank', return a boolean indicating
// whether or not the 'target' can be constructed by concatenating elements of 'wordBank'
// list.

// You may reuse elements of 'wordBank' as many times as needed.
package main

import (
	"fmt"
	"strings"
)

// Check if a given target can be constructed with a given word bank.
func CanConstruct(target string, wordBank []string, memo *map[string]bool) bool {
	if ans, ok := (*memo)[target]; ok {
		return ans
	}

	if target == "" {
		return true
	}

	for _, word := range wordBank {
		if strings.HasPrefix(target, word) {
			if CanConstruct(target[len(word):], wordBank, memo) {
				(*memo)[target] = true
				return true
			}
		}
	}

	(*memo)[target] = false
	return false
}

func main() {
	// Must create a map for each run of CanConstruct, because golang has no default value for
	// function params or optional params in function. If we pass the same map as param
	// to each run of CanConstruct, the values stored in the first run will affect the
	// subsequent ones because not all CanConstruct calls have the same targetSum. Another
	// solution would be to create just one map and after each run delete all results
	// stored inside it.
	memo1 := make(map[string]bool)
	memo2 := make(map[string]bool)
	memo3 := make(map[string]bool)
	memo4 := make(map[string]bool)

	fmt.Println(CanConstruct("abcdef", []string{"ab", "abc", "cd", "def", "abcd"}, &memo1))                                              // true
	fmt.Println(CanConstruct("skateboard", []string{"bo", "rd", "ate", "t", "ska", "sk", "boar"}, &memo2))                               // false
	fmt.Println(CanConstruct("enterapotentpot", []string{"a", "p", "ent", "enter", "ot", "o", "t"}, &memo3))                             // true
	fmt.Println(CanConstruct("eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef", []string{"e", "ee", "eee", "eeee", "eeeee", "eeeeee"}, &memo4)) // false
}
